\documentclass{beamer}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{cite}
\usepackage{tikz}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{tikzit}

\usepackage{xcolor}

\definecolor{myred}{RGB}{255, 0, 0}
\definecolor{myyellow1}{RGB}{255, 255, 219}
\definecolor{mygreen1}{RGB}{0, 255, 0}
\definecolor{mygreen2}{RGB}{0, 126, 0}
\definecolor{myblue}{RGB}{0, 0, 255}

\usetikzlibrary{arrows.meta,positioning}
\usetikzlibrary{shapes.multipart,matrix,positioning,arrows,arrows.meta}
\usefonttheme{serif}
\input{sample.tikzstyles}
% Edge styles

\setbeamertemplate{footline}[frame number]{}
\setbeamertemplate{navigation symbols}{}

\usecolortheme{lily}
\setbeamercolor{block title}{bg=blue!20,fg=black}
\setbeamercolor{block body}{bg = blue!10, fg = black}
\setbeamertemplate{itemize item}[square]
\setbeamercolor{itemize item}{fg = cyan}
\setbeamercolor{enumerate item}{fg = cyan}

\usetheme{default}
\beamertemplatenavigationsymbolsempty
\setbeamercolor{titlelike}{fg=blue}
\setbeamertemplate{caption}{\insertcaption\par}

%Information to be included in the title page:
\title{Characterizing evolutionary dynamics on a broader scale: a strain-space model for SARS-CoV-2}
\author{Peter C. Jentsch, PhD \inst{1,4} \and Finlay Maguire, PhD  \inst{3,5} \and Samira Mubareka, MD, FRCPC \inst{1,2}}
\institute{\inst{1} Sunnybrook Research Institute, Toronto, Canada  \and \inst{2} University of Toronto, Toronto, Canada \and \inst{3} Dalhousie University, Halifax, Canada \and \inst{4} Simon Fraser University, Burnaby, Canada \and \inst{5} Shared Hospital Laboratory, Toronto, Canada}
\date{\today}

\begin{document}
\setbeamertemplate{caption}{\insertcaption\par}
\frame{\titlepage}

\begin{frame}{Infection spread with compartmental models}
    % \begin{itemize}
    %     \item Infection spread is often modelled using compartmental models
    %     \item Represent subsets of a host population and rates of movement between them
    % \end{itemize}
    \begin{figure}
        \centering
            \begin{tikzpicture}[node distance=2.5cm, auto,
                >=Latex, 
                every node/.append style={align=center},
                int/.style={draw, minimum size=1.5cm}]
            
            \node [int] (S)             {$S$};
            \node [int, right=of S ] (I) {$I$};
            \node [int, right=of I] (R) {$R$};
            \path[->] (S) edge node {$\beta S I$} (I);
            \path[->] (I) edge node {$\gamma I$} (R);
            \end{tikzpicture}
            \label{SIR_diagram}
    \end{figure}
    \begin{itemize}
        \item $S(t)$ - fraction of susceptible hosts at $t$
        \item $I(t)$ - fraction of infected hosts at $t$
        \item $R(t)$ - fraction of recovered hosts at $t$
        \item $1 = S(t) + I(t) + R(t)$ is constant
    \end{itemize} 
\end{frame}
\begin{frame}{Broad use of genomic data in dynamical epidemic models has been limited}
    \begin{figure}
        \centering
        \includegraphics[width=0.8\textwidth]{2022-07-12-00-57-38.png}
    \end{figure}
    \centering
    \vfill
    \tiny{\cite{grenfell2004unifying}}
\end{frame}
\begin{frame}{Modeling multiple infections is challenging}
    % \begin{itemize}
    %     \item Infection spread is often modelled using compartmental models
    %     \item Represent subsets of a host population and rates of movement between them
    % \end{itemize}
    \begin{figure}
        \centering
            \includegraphics[width=0.8\textwidth]{2022-07-10-22-37-45.png}
    \end{figure}
    \centering
    \vfill
    \tiny{\cite{Alizon_van_Baalen_2008}}
    
\end{frame}

\begin{frame}{Complexity only increases with vaccination dynamics etc.}
% \begin{columns}
%         \begin{column}{0.6\textwidth}

%             \begin{itemize}
%                 \item Multiple infections (e.g. competing VoCs) can be represented as more compartments
%                 \item Work on multiple infections is usually here due to lack of data, increasing complexity
%             \end{itemize}
    
% \end{column}
\begin{figure}[h!]
    \centering
    \scalebox{0.5}{
    \input{diagram_wide.tikz}
    }
% \caption{State diagram of model with two different types of vaccination and strong competition.}
\label{model_structure}
\end{figure}

%     \begin{column}{\textwidth}

% \begin{figure}[h!]
%     \centering
%     \includegraphics[width=\textwidth]{voc_model_diagram.png}
% \label{model_structure}
% % \end{figure}
% \end{column}
% \end{columns}
% (e.g. \cite{Alizon_van_Baalen_2008, van_Baalen_Sabelis_1995, Lipsitch_Colijn_Cohen_Hanage_Fraser_2009, Nicoli_Ayabina_Trotter_Turner_Colijn_2015})
\end{frame}

\begin{frame}{How to incorporate more data?}
    \centering
    \begin{figure}
        \includegraphics[width=0.9\textwidth]{my_figs/phylogeny.png}
    \end{figure}
\end{frame}
\begin{frame}{A useful approximation}
    \begin{columns}
\begin{column}{\textwidth}
    \begin{figure}
        \includegraphics[width=0.8\textwidth]{standalone/gog_paper.png}
    \end{figure}
    \begin{figure}
        \includegraphics[width=0.8\textwidth]{standalone/gog_paper_2.png}
    \end{figure}
\end{column}
\end{columns}
\end{frame}
    %     \item Can extend these models to a sequence of variants
    %     \item Assume each variant is indexed by $i$ 
    %     \item The dynamics at each variant $i$ are determined by a simple compartmental model
            % \begin{column}{0.5\textwidth}
        %     \begin{itemize}
        %         \item Variants are related by a function $\sigma(i,j)$ that determines how much an infection by variant $i$ reduces probability of infection to variant $j$.
        %         \item A variant $i$ mutates to neighbouring indices $i+1, i-1$ proportional to the population of variant $i$
        %         \end{itemize}         
        % \end{column}

\begin{frame}{Begin with SIR model}
    \begin{columns}       
         \begin{column}{0.5\textwidth}
        \begin{itemize}
            \item Consider strains of a pathogen numbered $i \in [1,N]$ 
            \item $S(t)$ - population of hosts susceptible to strain $i$ at $t$
            \item $I(t)$ - population of hosts infected with strain $i$ at $t$
            \item $R(t)$ - population of hosts recovered from strain $i$ at $t$
        \end{itemize} 
    \end{column}
        \begin{column}{0.5\textwidth}
            \begin{figure}
                \centering
                \scalebox{1.0}{
                \input{tikz_models/gog_model_1.tikz}
                }
            \end{figure}
        \end{column}
        \end{columns}
\end{frame}
\begin{frame}{Organize strains on a lattice}
    \begin{figure}
        \centering
        \scalebox{0.8}{
        \input{tikz_models/gog_model_2.tikz}
        }
    \end{figure}
\end{frame}
\begin{frame}{Strains mutate into adjacent strains}
    \begin{figure}
        \centering
        \scalebox{0.8}{
        \input{tikz_models/gog_model_3.tikz}
        }
    \end{figure}
\end{frame}
\begin{frame}{Infections provide cross-immunity to nearby strains}
        \centering
            \begin{figure}
                \includegraphics[width=0.6\textwidth]{standalone/gog_fig.png}
            \end{figure}
    \centering
    \vfill
    \tiny{\cite{gogDynamicsSelectionManystrain2002}}
\end{frame}
%notice that the flow of population out of each susceptible compartment $i$ is the number of people no longer susceptible to strain $i$, and only a proportion of those go into the infected compartment i
\begin{frame}{Full strain lattice model in 1 dimension}
    \begin{figure}
        \centering
        \scalebox{0.8}{
        \input{tikz_models/gog_model_4.tikz}
        }
    \end{figure}
\end{frame}
        %         \item Variants are related by a function $\sigma(i,j)$ that determines how much an infection by variant $i$ reduces probability of infection to variant $j$. Note that sigma(i,i) = 1.
\begin{frame}{Extending the strain lattice to 2 dimensions}
    \begin{columns}
        \begin{column}{0.5\textwidth}
            \begin{figure}
                \includegraphics[width=\textwidth]{gog_2d_fig/gog_2d_a_1.png}
            \end{figure}        
        \end{column}
        \begin{column}{0.5\textwidth}
        \end{column}
    \end{columns}
    \centering
    \vfill
    \tiny{\cite{gogDynamicsSelectionManystrain2002}}

\end{frame}
\begin{frame}{Extending the strain lattice to 2 dimensions}
    \begin{columns}
        \begin{column}{0.5\textwidth}
            \begin{figure}
                \includegraphics[width=\textwidth]{gog_2d_fig/gog_2d_a.png}
            \end{figure}        
        \end{column}
        \begin{column}{0.5\textwidth}
            \begin{figure}
                \includegraphics[width=\textwidth]{gog_2d_fig/gog_2d_b.png}
            \end{figure}
        \end{column}
    \end{columns}
    \centering
    \vfill
    \tiny{\cite{gogDynamicsSelectionManystrain2002}}

\end{frame}
%This is the abstraction I will use to make it relatively tractable to model a lattice of related infection strains, complexity scales linearly
% As implied above, we use “polarized immunity,” which means that partial cross-immunity acts to render some of the hosts totally immune. The partial aspect of the immunity is introduced through a probability that a given host will gain immunity; only some hosts gain immunity when exposed to a partially crossreacting strain; other hosts do not gain immunity from that exposure. 
%The next step is to place at least some of the genomic data on the, hopefully 2 or 3 dimensional, lattice.
%in order to do this, we need some way to quantify the differences in immune response to viruses.

%Most methods of assessing immune response involve taking serum that has been raised against one strain, antiserum, and exposing it to another strain to determine how many antibodies bind, which is done in an neutralization assay.

\begin{frame}{Quantifying antigenicity between viruses}
    \begin{figure}
        \centering
        \scalebox{0.9}{
        \input{tikz_models/antigenic_fig.tikz}
        }
    \end{figure}
\end{frame}

% The same serum sample is spread among well plates, progressively more diluted in each well. A sample of virus with known concentration is added to each well, and a test is conducted to see if all the virus has been bound with antibodies. Hemagglutinating viruses, such as influenza, are called this because they bind to red blood cells preventing them from clumping. Since they cannot bind to red blood cells if they're bound to antibodies, antibodies binding can be assessed this way. This method determines the titre of antibodies in the serum sample that react to the viral particles.
% SARS-CoV-2 neutralizing assays work similarly, except using dyes to assess antibody binding and outcomes are often similar.

% Even a small number of viral samples and corresponding antisera will product a very high-dimensional dataset, which is not amenable to the 2 or 3 dimensional space that we want to use.
% Therefore, Alan Lapedes and Robert Farber came up with the used a modification of a technique called Multidimensional scaling, that given distances, produces a lower-dimensional manifold that minimizes the changes to the distances. 


\begin{frame}{Projecting points to low dimensions while preserving distances}
    \begin{columns}
        \begin{column}{0.4\textwidth}
            \begin{figure}
                \includegraphics[width=1.4\textwidth]{standalone/mds_1.png}
                
            \end{figure}   
        \end{column}
        \begin{column}{0.01\textwidth}
            \huge{$\Longrightarrow$}
            
        \end{column}
        \begin{column}{0.4\textwidth}
        %     \begin{figure}
        %     \includegraphics[width=\textwidth]{standalone/mds_2.png}
            
        % \end{figure}   
        \end{column}
    \end{columns}
    \centering
    \vfill
    \tiny{\cite{scikit-learn}}
\end{frame}
\begin{frame}{Projecting to low dimensions while preserving distances}
    \begin{columns}
        \begin{column}{0.4\textwidth}
            \begin{figure}
                \includegraphics[width=1.4\textwidth]{standalone/mds_1.png}
                
            \end{figure}   
        \end{column}
        \begin{column}{0.01\textwidth}
            \huge{$\Longrightarrow$}
            
        \end{column}
        \begin{column}{0.4\textwidth}
            \begin{figure}
            \includegraphics[width=\textwidth]{standalone/mds_2.png}
            
        \end{figure}   
        \end{column}
    \end{columns}
    \centering
    \vfill
    \tiny{\cite{scikit-learn}}

\end{frame}
% \begin{frame}{Approximating genomic data in 2 dimesions }

%     \begin{columns}
%         \begin{column}{0.6\textwidth}
%         \begin{figure}
%         \includegraphics[width=\textwidth]{influenza_map/smith_antigenic_1.png}    
%         \end{figure}   
%     \end{column}
%     \end{columns}
%     \centering
%     \vfill
%     \tiny{\cite{lapedesGeometryShapeSpace2001,smithMappingAntigenicGenetic2004}}
%     % \begin{itemize}
%     %     \item Practice of mapping out immune responses to related pathogens
%     %     \item Distance between serums and pathogen is quantified, these points are visualized using multidimensional scaling (MDS)
% % ,  usually as proportional reduction in serum viral titre hemaagglutinin assay
%     % \end{itemize}
% \end{frame}

%This technique was developed for influenza 

\begin{frame}{Approximating genomic data in 2 dimesions }

    \begin{columns}
        \begin{column}{0.4\textwidth}
        \begin{figure}
        \includegraphics[width=\textwidth]{influenza_map/smith_antigenic_2.png}    
        \end{figure}   
    \end{column}
    \end{columns}
    \centering
    \vfill
    \tiny{\cite{lapedesGeometryShapeSpace2001,smithMappingAntigenicGenetic2004}}
    % \begin{itemize}
    %     \item Practice of mapping out immune responses to related pathogens
    %     \item Distance between serums and pathogen is quantified, these points are visualized using multidimensional scaling (MDS)
% ,  usually as proportional reduction in serum viral titre hemaagglutinin assay
    % \end{itemize}
\end{frame}


\begin{frame}{Mapping SARS-CoV-2 in 2D}
        \begin{figure}
        \includegraphics[width=\textwidth]{sars_cov_2_map/wilks_1.png}
    \end{figure}   
    \centering
    \vfill
    \tiny{\cite{wilksMappingSARSCoV2Antigenic2022}}
\end{frame}

%wilks et al are using geometric mean titre here 
\begin{frame}{Mapping SARS-CoV-2 in 2D}
    \begin{figure}
    \includegraphics[width=\textwidth]{sars_cov_2_map/wilks_3.png}
\end{figure}   
\centering
\vfill
\tiny{\cite{wilksMappingSARSCoV2Antigenic2022}}
\end{frame}

%how to interpolate more data into this map?
%start with RBD stuff
% \begin{itemize}
%     \item A huge amount of genomic data is available
%     % \item We have millions of samples and hundreds of thousands of unique genomes from random sampling alone
%     \item Assign each sample to its closest lineage in the map from \cite{wilksMappingSARSCoV2Antigenic2022}
%     \item Two methods to further differentiate genomes: homoplasic mutations or antibody binding
%     % Distance between any two samples is determined by the distance between the respective closest lineage in Wilks, plus the SNP distance, where each SNP is weighted by how often it recurs in the global tree
% \end{itemize}


\begin{frame}{Methods of adding more genomes: polyclonal antibody binding studies}
    \begin{figure}
        \includegraphics[width=\textwidth]{standalone/2022-07-06-13-06-32.png}
    \end{figure}
    \centering
\vfill
\tiny{\cite{greaney2022antibody}}
\end{frame}

\begin{frame}{Interpolating more data into antigenic map}
    For each pair of genomes $g_i$ and $g_j$ ($i\neq j$)
    \begin{enumerate}
        \item Find the closest lineages in existing antigenic map to $g_i$, and $g_j$, and corresponding points $x_i$ and $x_j$
        \item 
        \item
        \item 
    \end{enumerate}
\end{frame}
\begin{frame}{Interpolating more data into antigenic map}
    For each pair of genomes $g_i$ and $g_j$ ($i\neq j$)
    \begin{enumerate}
        \item Find the closest lineages in existing antigenic map to $g_i$, and $g_j$, and corresponding points $x_i$ and $x_j$
        \item Determine difference in polyclonal binding affinity $B(g_i,g_j)$
        \item
        \item
    \end{enumerate}
\end{frame}
\begin{frame}{Interpolating more data into antigenic map}
    For each pair of genomes $g_i$ and $g_j$ ($i\neq j$)
    \begin{enumerate}
        \item Find the closest lineages in existing antigenic map to $g_i$, and $g_j$, and corresponding points $x_i$ and $x_j$
        \item Determine difference in polyclonal binding affinity $B(g_i,g_j)$
        \item Distance between $g_i$ and $g_j$ is $||x_i - x_j|| + B(g_i,g_j)$
        \item
    \end{enumerate}
\end{frame}
\begin{frame}{Interpolating more data into antigenic map}
    For each pair of genomes $g_i$ and $g_j$ ($i\neq j$)
    \begin{enumerate}
        \item Find the closest lineages in existing antigenic map to $g_i$, and $g_j$, and corresponding points $x_i$ and $x_j$
        \item Determine difference in polyclonal binding affinity $B(g_i,g_j)$
        \item Distance between $g_i$ and $g_j$ is $||x_i - x_j|| + B(g_i,g_j)$
        \item Use these distances as inputs to multidimensional scaling
    \end{enumerate}
\end{frame}
\begin{frame}{Antibody Binding map}
    \begin{figure}
        \centering
        \includegraphics[width=1.1\textwidth]{my_figs/binding_usa_multidimensional_scaling.png}
    \end{figure}
\end{frame}

\begin{frame}{Methods of adding more genomes: homoplasic sites}
    \begin{figure}
        \centering
        \scalebox{1.1}{
        \input{tikz_models/homoplasy.tikz}
        }
    \end{figure}
    \centering
    \vfill
    \tiny{\cite{page2009molecular}}
\end{frame}

%unconstrained site
%systematic error in sequencing 
\begin{frame}{Homoplasic mutations map}
    \begin{figure}
        \centering

        \includegraphics[width=1.1\textwidth]{my_figs/homoplasy_usa_multidimensional_scaling.png}
    \end{figure}
\end{frame}


\begin{frame}{Evaluating the MDS approximation}
    \begin{figure}
        \includegraphics[width=\textwidth]{my_figs/usa_mds_stress.png}
    \end{figure}
\end{frame}

    
% \begin{frame}{Kernel approximations of}
%     \begin{figure}
%         \includegraphics[width=\textwidth]{my_figs/kernel_approx_still.png}
%     \end{figure}
% \end{frame}
\begin{frame}{Model parameters/variables}
    \begin{table}[h!]
        \begin{center}
        \begin{tabular}{c|p{8cm}}
                Symbol & Description\\
                \hline
                \hline
                $N$ & Size of variant grid \\
                $K$ & Total population \\
                $S_{ij}$ & Population susceptible to variant $(i,j) \in [0,N]^2$ \\
                $I_{ij}$ & Population infected by variant $(i,j) \in [0,N]^2$\\
                $R_{ij}$ & Recovered/Immune to variant $(i,j) \in [0,N]^2$\\
                $\sigma_{ijkl}$ & Probability that exposure to variant $(i,j)$ causes immunity \newline to variant $(k,l)$\\
                $\beta_{ij}$ & Transmission rate of variant $(i,j)$\\
                $v(t)$ & vaccination rate at time $t$\\
                $s(t)$ & stringency at $t$\\
                $\xi$ & Recovery rate of all strains \\
                $\gamma$ & Rate of immunity loss of all strains \\
        \end{tabular}
        \caption{Table of symbols for Model 2}
    
        \label{variables_2}
        \end{center}
    \end{table}
\end{frame}
\begin{frame}{Model Equations}
    \small
    \begin{equation}
        \frac{S_{ij}}{dt} = \textcolor{myred}{-\frac{1}{K}\sum_{kl} s(t) \beta_{kl} \sigma_{ijkl} S_{ij} I_{kl}} + \gamma (K - S_{ij} - I_{ij}) - V(t)S  \label{Seqn}
    \end{equation}
    \begin{equation}
        \frac{ I_{ij}(t)}{dt} = \textcolor{myred}{\frac{1}{K}s(t) \beta_{ij} S_{ij} I_{ij}} - \xi I_{ij} + \textcolor{myblue}{M \left(- 4I_{ij} + I_{i-1,j}  + I_{i+1,j} + I_{i,j-1} + I_{i,j+1} \right)} \label{Ieqn}    
    \end{equation}
    % \begin{equation}
    %     \frac{R_{ij}(t)}{dt} = \xi I_{ij} - \gamma R_{ij} + V(t)S \label{Reqn}
    % \end{equation}

    Boundary conditions: $I_{0,j} = 0, I_{j,0} = 0,  I_{N,j} = 0, I_{j,N} = 0$
\end{frame}
\begin{frame}{Model Equations}
    \small
    \begin{equation}
        \frac{S_{ij}}{dt} = \textcolor{myred}{-\frac{1}{K}\sum_{kl} s(t) \beta_{kl} \sigma_{ijkl} S_{ij} I_{kl}} + \gamma R - V(t)S  \label{Seqn}
    \end{equation}
    \begin{equation}
        \frac{ I_{ij}(t)}{dt} = \textcolor{myred}{\frac{1}{K}s(t) \beta_{ij} S_{ij} I_{ij}} - \xi I_{ij} + \textcolor{myblue}{M \left(- 4I_{ij} + I_{i-1,j}  + I_{i+1,j} + I_{i,j-1} + I_{i,j+1} \right)} \label{Ieqn}    
    \end{equation}
    \begin{equation}
        \frac{R_{ij}(t)}{dt} =  \textcolor{myred}{\frac{1}{K}\sum_{kl \neq ij} s(t) \beta_{kl} \sigma_{ijkl} S_{ij} I_{kl}} + \xi I_{ij} - \gamma R_{ij} + V(t)S \label{Reqn}
    \end{equation}

    Boundary conditions: $I_{0,j} = 0, I_{j,0} = 0,  I_{N,j} = 0, I_{j,N} = 0$
\end{frame}
% \begin{frame}
% \begin{columns}
% \begin{column}{\textwidth}
%     \begin{itemize}
%         \item Dynamic model of Sars-CoV-2 evolution, representing antigenic diversity on a lattice
%         %  (as in e.g. \cite{gogDynamicsSelectionManystrain2002,kryazhimskiyStateSpaceReductionMultiStrain2007, Marchi_2021})
%         \item Antigenically distinct variants of the virus are mapped to 2D grid,  distance between variants corresponds to the proportional reduction in maximum serum viral titre 
    
%         % \cite{wilksMappingSARSCoV2Antigenic2022,van2022mapping}
%     \end{itemize}
% \begin{figure}
%     \includegraphics[width=0.5\textwidth]{../SarsEvoModel/plots/antigenic_map_paper.png}
%     \caption{\small{reproduced from \cite{wilksMappingSARSCoV2Antigenic2022}, Fig. 2}}
% \end{figure}
% \end{column}

% \end{columns}
% \end{frame}


\begin{frame}{Model Dynamics}
    \begin{figure} 
        \includegraphics[width=1.1\textwidth]{2022-07-12-01-13-30.png}
    \end{figure}
\end{frame}


\begin{frame}{Further work}
    \begin{itemize}
        \item More detailed methods for interpolation of additional genome data
        \item Differentiate between antigenic and genomic space
        \begin{itemize}
            \item[$\rightarrow$] non-local diffusion
        \end{itemize}
        \item Model fitting
        \item Applications
        \begin{itemize}
            \item Simple inference on antigenic space
            \item VoC-aware NPI usage and vaccination
        \end{itemize}
    \end{itemize}
\end{frame}

% \begin{frame}{}
%     \small
%     To incorporate more realistic mutation rates, we can go to continuous strain-space and use nonlocal reaction-diffusion dynamics as in \cite{Rouzine_Rozhnova_2018,Bessonov_Bocharov_Meyerhans_Popov_Volpert_2021}
%     \tiny
%     \begin{equation}
%         S_t(x,y,t) = \int_{-\infty}^{\infty} \int_{-\infty}^{\infty} \beta(x',y') \sigma(x,y,x', y') S(x,y,t) I(x',y',t) dx' dy' + \gamma R_{ij} -  \eta(t) v(x,y) S(x,y,t)\label{Seqn_cts}
%     \end{equation}
%     \begin{equation}
%         I_t(x,y,t) = \beta(x,y) S(x,y,t) I(x,y,t)- \xi I(x,y,t) + M \left(I_x(x,y,t)  + I_y(x,y,t)  \right) \label{Ieqn_cts}    
%     \end{equation}
%     \begin{equation}
%         R_t(x,y,t) = \xi I(x,y,t)I(x,y,t) - \gamma R(x,y,t) + \eta(t) v(x,y) S(x,y,t) \label{Reqn_cts}
%     \end{equation}
    
%     where $\beta, \sigma, v$ have been generalized to their continuous counterparts. Given a dispersion kernel $K(x,y) \in L_2: \mathbb{R}^2 \to \mathbb{R}$ this can be generalised to non-local diffusion as follows
    
%     \begin{equation}
%         I_t(x,y,t) = \beta(x,y) S(x,y,t) I(x,y,t)- \xi I(x,y,t) + M \left(\int_{-\infty}^{\infty} \int_{-\infty}^{\infty} K(x-x',y-y')I(x',y',t) dx' dy' \right) \label{Ieqn_cts_nonlocal}    
%     \end{equation}
% \end{frame}
% \begin{frame}{Developing an antigenic distance map}
%     \begin{itemize}
%         \item We would like an approximate measure of antigenic distance for every sample genome
%         \item Using all samples, we compute pairwise distances between each unique genome in some way that encodes antigenic response   
%         \item Many possible ways to do this, so far none of them seem to work very well
%         \item Project to 2-d (hopefully) space with multidimensional scaling
%     \end{itemize}    
% \end{frame}
% \begin{frame}{Genome distance}
%     Assume:
%     \begin{itemize}
%         \item $a,b$ are SARS-CoV-2 genomes aligned with the reference
%         \item $a_i$ the $i$th nucleotide base in $a$ and
%          \[
%             \chi(a_i,b_i) =   \begin{cases}
%             1 & \text{if $a_i=b_i$} \\
%             0 & \text{otherwise} \\
%         \end{cases}
%         \]
%         \item $h_i$ is a vector containing the number of homoplasic mutations at site $i$ in the global tree
%         \item $\mathfrak{B}(a)$ computes the polyclonal binding affinity of genome $a$ as per \cite{starr2020deep}
%     \end{itemize}
        
%     One option for a distance measure is something like
%     \begin{equation}
%         d(a,b) = \frac{\mathfrak{B}(a) + \mathfrak{B}(b)}{2} + \sum_i \chi(a_i,b_i) h_i
%     \end{equation}
%     That is, the average binding between two genomes plus the SNP distance weighted by the relative homoplasy of each mutation.
% \end{frame}
% \begin{frame}{Example antigenic distance map}
%     \begin{figure}
%         \includegraphics[width=\textwidth]{uk_multidimensional_scaling.png}
%         \caption{Multidimensional scaling plot using samples from the UK up to mid November}
%     \end{figure}
% \end{frame}
% \begin{frame}{Homoplasy in global tree}
%     \begin{figure}
%         \includegraphics[width=\textwidth]{gene_freq.png}
%         \caption{Number of recurrent (homoplasic) mutations per base by gene, (normalized by gene length)}
%     \end{figure}
% \end{frame}
% \begin{frame}{Homoplasy in orf3a}
%     \begin{figure}
%         \includegraphics[width=\textwidth]{orf3a_scatter_2.png}
%     \end{figure}
% \end{frame}


\begin{frame}[allowframebreaks]

\bibliographystyle{apalike}
\bibliography{ref.bib}
\end{frame}
\begin{frame}{Mutation homoplasy}
    \begin{figure}
        \includegraphics[width=\textwidth]{my_figs/gene_freq.png}
    \end{figure}
\end{frame}
\end{document}