\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{tikz}
\usepackage{tikzit}

\usepackage{authblk}
\usepackage{blindtext}

\usepackage{url}
\usepackage{cite}
%% -------------------------------------- Declare the layers
\pgfdeclarelayer{nodelayer}
\pgfdeclarelayer{edgelayer}
\pgfsetlayers{edgelayer,nodelayer,main}
% Node styles
\tikzstyle{black circle}=[fill=white, draw=black, shape=circle, tikzit shape=circle]
\tikzstyle{Arrow}=[->]

% Edge styles
\tikzstyle{new edge style 0}=[->]

\title{Characterizing evolution pressures in Sars-CoV-2 on a broader scale}
\author[1,4]{Peter C. Jentsch}  
\author[3,5]{Finlay Maguire}
\author[1,2]{Samira Mubareka}
\affil[1]{Sunnybrook Research Institute, Toronto, Canada}
\affil[2]{University of Toronto, Toronto, Canada}
\affil[3]{Dalhousie University, Halifax, Canada}
\affil[4]{Simon Fraser University, Burnaby, Canada}
\affil[5]{Shared Hospital Laboratory, Toronto, Canada}
\date{\today}                     %% if you don't need date to appear
\setcounter{Maxaffil}{0}
\renewcommand\Affilfont{\itshape\small}

\begin{document}


\maketitle

\section{Introduction}

(Peter: need to rework much of this text...)


The objective of this work is to study the SARS-CoV-2 viral diversity in Ontario and Quebec. Initially, the focus will be comparing differences between the provinces. The prevalence of the Gamma and Alpha variants of concern (VOC) as differed significantly between Ontario and Quebec during the 3rd wave of the pandemic in Canada (February 2021 to June 2021). The key differences between Alpha and Gamma are in immune escape and transmissibility \cite{agency_for_clinical_innovation_2022}. Alpha has less immune escape (for neutralising sera from vaccinated or convalescent individuals), but binds ACE2 more readily and therefore relatively higher transmission. We will not need to consider the S:E484k mutation, which affects immune escape, since it was in a negligible proportion of Canadian Alpha infections during this time period. On the other hand, the Gamma lineage likely has greater immune escape but is probably relatively less transmissible. In contrast to Ontario, during this time the government of Quebec was actively prioritizing the distribution of mRNA vaccines to groups more vulnerable to hospitalization and communities with VOC outbreaks. Therefore, both the mRNA (Pfizer, Moderna) and AstraZeneca vaccine will be included in our model.

Compartmental differential equation models have been used extensively to model competing viral strains in the past (e.g. \cite{Alizon_van_Baalen_2008, van_Baalen_Sabelis_1995, Lipsitch_Colijn_Cohen_Hanage_Fraser_2009, Nicoli_Ayabina_Trotter_Turner_Colijn_2015}), and have been invaluable to informing policy throughout the pandemic \cite{Mishra_Berah_Mellan_Unwin_Vollmer_Parag_Gandy_Flaxman_Bhatt_2020,jentsch2021prioritising, bubar2021model}.  One approach to address this research question is to develop a parsimonious model that accurately reproduces observed data. Such a model would generate generalizable insights into the effect of NPIs more broadly. To this end, here we suggest a compartmental model to act as a jumping-off point (Section \ref{model}, Figure \ref{model_structure}), and request data fields to parameterize it (Section \ref{data_fields}). Most disease parameters can be obtained from literature. We will estimate underreporting and underascertainment multipliers by comparing observed case data with serological studies \cite{jentsch2021prioritising}.


\section{Model}

The evolution of Sars-CoV-2 throughout the pandemic is marked by antigenic drift \cite{yewdellAntigenicDriftUnderstanding2021}, giving rise to new variants that exhibit significant immune escape. The model of Gog and Grenfell \cite{gogDynamicsSelectionManystrain2002} constrains strain space to a one or two dimensional lattice, thereby making the analysis of strain evolution tractable. In two dimensional strain space, cross-immunity of a pathogen is given specified by a coefficient $\sigma_{ijkl}$, and mutation is implemented as discrete diffusion with some fixed speed. These ideas were generalized to n-dimensional strain space, and applied to modeling drift in influenza A by Kryazhimskiy et al \cite{kryazhimskiyStateSpaceReductionMultiStrain2007}. This application of these models to influenza was partially motivated by the work of Lapedes and Farber \cite{lapedesGeometryShapeSpace2001}, and Smith et al. \cite{smithMappingAntigenicGenetic2004}, which suggests that the antigenic evolution of Influenza A primarily occurs within two dimensions.
The idea for this line of Sars-CoV-2 research is to handle strain space similar to the aforementioned work, but include further compartments and dynamics specific to Sars-CoV-2 to test hypotheses of evolution. 


\begin{equation}
    S_{ij}'(t) = -\sum_{kl} \beta_{kl} \sigma_{ijkl} S_{ij} I_{kl} + \gamma R_{ij}  \label{Seqn}
\end{equation}
\begin{equation}
    I_{ij}'(t) = \beta_{ij} S_{ij} I_{ij} - \xi I_{ij} + M \left(- 4I_{ij} + I_{i-1,j}  + I_{i+1,j} + I_{i,j-1} + I_{i,j+1} \right) \label{Ieqn}    
\end{equation}
\begin{equation}
    R_{ij}'(t) = \xi I_{ij} - \gamma R_{ij}  \label{Reqn}
\end{equation}


\begin{table}[h!]
    \begin{center}
        \begin{tabular}{c|p{8cm}}
            Symbol & Description \\
            \hline
            \hline
            $N$ & Size of variant grid \\
            $S_{ij}$ & Population susceptible to variant $(i,j) \in [0,N]^2$ \\
            $I_{ij}$ & Population infected by variant $(i,j) \in [0,N]^2$\\
            $R_{ij}$ & Recovered/Immune to variant $(i,j) \in [0,N]^2$\\
            $\sigma_{ijkl}$ & Probability that exposure to variant $(i,j)$ causes immunity \newline to variant $(k,l)$\\
            $\beta_{ij}$ & Transmission rate of variant $(i,j)$\\
            $\xi$ & Recovery rate of all strains \\
            $\gamma$ & Rate of immunity loss of all strains \\
    \end{tabular}
    \caption{Table of symbols for Model 2}

    \label{variables_2}
    \end{center}
\end{table}

    

Equations \ref{Seqn}-\ref{Reqn} represent the model of \cite{gogDynamicsSelectionManystrain2002} with a two-dimensional strain space, with a few changes to better reflect mechanisms of of Sars-CoV-2. I have added an immunity period, changing the model to an SIRS mechanism, and removed the vital dynamics, as I do not think natural population birth rates are significant in the time scale of the pandemic. A key assumption made by this model is that exposure grants complete immunity to some fraction of individuals, rather than partial immunity (interpreted as reduced transmission rates) to all exposed individuals. Many other methods of dealing with cross-immunity are possible, but this method gives a simpler state space \cite{Castillo_Chavez_Blower_Driessche_Kirschner_Yakubu_2002}.


To incorporate vaccination, consider each vaccine affecting a different region of strain space. That is, for a vaccine $v$ we can associate a matrix $v_{ij} \in [0,1]$ which determines the relative effect of that vaccine on the immunity of hosts to strain $i,j$. The function $eta(t)$ represents some base rate of vaccination. This results in the following equations for $S_{ij}'(t)$ and $ R_{ij}'(t) $ (the infected equation \ref{Ieqn} is unchanged).


\begin{equation}
    S_{ij}'(t) = -\sum_{kl} \beta_{kl} \sigma_{ijkl} S_{ij} I_{kl} + \gamma R_{ij} -  \eta(t) v_{ij} S_{ij} \label{Seqn}
\end{equation}
\begin{equation}
    R_{ij}'(t) = \xi I_{ij} - \gamma R_{ij} + \eta(t) v_{ij} S_{ij} \label{Reqn}
\end{equation}


This model could be used to test the effect of NPIs on Sars-CoV-2 evolution. Mechanisms for NPIs and additional compartments for heterogeneity are straightforward to add to the model equations. If a given NPI mechanism is more able to fit data with a mechanism that does not act on all strains equally, then this could be evidence that NPIs are affecting Sars-CoV-2 evolution. 

\subsection{Continuous strain-space}

The above model can be viewed as simply a first-order finite difference approximation of a continuous space reaction-diffusion model. Accordingly, we can generalize it to continuous strain-space as
 
\begin{equation}
    S_t(x,y,t) = \int_{-\infty}^{\infty} \int_{-\infty}^{\infty} \beta(x',y') \sigma(x,y,x', y') S(x,y,t) I(x',y',t) dx' dy' + \gamma R_{ij} -  \eta(t) v(x,y) S(x,y,t)\label{Seqn_cts}
\end{equation}
\begin{equation}
    I_t(x,y,t) = \beta(x,y) S(x,y,t) I(x,y,t)- \xi I(x,y,t) + M \left(I_x(x,y,t)  + I_y(x,y,t)  \right) \label{Ieqn_cts}    
\end{equation}
\begin{equation}
    R_t(x,y,t) = \xi I(x,y,t)I(x,y,t) - \gamma R(x,y,t) + \eta(t) v(x,y) S(x,y,t) \label{Reqn_cts}
\end{equation}

where $\beta, \sigma, v$ have been generalized to their continuous counterparts. This formulation is similar to the 1-dimension strain space model described in \cite{Bessonov_Bocharov_Meyerhans_Popov_Volpert_2021}. Then, given a dispersion kernel $K(x,y) \in L_2: \mathbb{R}^2 \to \mathbb{R}$ this can be generalised to non-local diffusion as follows

\begin{equation}
    I_t(x,y,t) = \beta(x,y) S(x,y,t) I(x,y,t)- \xi I(x,y,t) + M \left(\int_{-\infty}^{\infty} \int_{-\infty}^{\infty} K(x-x',y-y')I(x',y',t) dx' dy' \right) \label{Ieqn_cts_nonlocal}    
\end{equation}


\begin{figure}[h!]
    \centering
    \input{diagram.tikz}
\caption{State diagram of model with two different types of vaccination and strong competition. Variables are described in Table \ref{variables}}
\label{model_structure}
\end{figure}

\subsection{Parameterization}

Previous work on these models do not use data to estimate parameters, focusing on broad characterization of dynamics with numerical or analytical approaches. However, genomic data does include a huge amount of admittedly very noisy information. There is some recent work on parameter estimation by comparing simulated phylogeny with observed phylogeny using a suite of summary statistics for tree structures \cite{danesh2021quantifying,leventhal2012inferring,saulnier2017inferring}, but these models do not explicitly include genomic structure, which might provide additional inferential power. 

Placing individual samples into the collapsed strain space could be done by using neutralizing antibody studies. The neutralizing response has been characterized for most variants of concern with respect to monoclonal antibodies, convalescent plasma, and plasma from vaccinated persons \cite{stanford2020}. To an even higher level of detail. Starr et al. characterized polyclonal antibody binding over all mutations of the Receptor Binding Domain (RBD) \cite{starr2020deep}. If the neutralizing response is known, we assign strains positions such that the distance between any two strains is equal to that neutralizing response. If the response is not known, we just assume that the position is close to the most recent ancestor for which the response is known (likely a VoC). This approach is especially appropriate for recurrent mutations, since it is unlikely that a mutation that has occurred many times in the viral history but has not evolved further confers a significant advantage. The immune-space position of vaccinate-induced immunity can also be computed this way, although we will still need to make significant assumptions on their geometry. It might be best to assume a simple symmetric shape with a centre given by neutralization results. 

Mirroring the work on mapping the strain space for influenza \cite{lapedesGeometryShapeSpace2001, smithMappingAntigenicGenetic2004, cai2010computational} computing this embedding is a metric multidimensional scaling problem. This method has been applied to the antigenic space of the Sars-CoV-2 as well \cite{millerAntigenicSpaceFramework2021, wilksMappingSARSCoV2Antigenic2022, van2022mapping}. 

\bibliographystyle{plain}
\bibliography{ref.bib}

\end{document}  


% \section{Model}
% \label{model}
% \begin{itemize}
%     \item The structure of the model is Susceptible-Exposed-Infected-Asymptomatic-Recovered, outlined in Figure \ref{model_structure}.
%     \item The mRNA vaccine lowers transmission rates and increases the proportion of asymptomatic infections compared to the AstraZeneca vaccine
%     \item Focus on infections in two lineages
%     \item Each lineage has different infection rates, recovery times, and latent periods
%     \item Co-infection negligible 
%     \item Time scale short enough that recovery from either infection grants immunity
%     \item Strains interact in the model only in that they compete for hosts
%     \item Vaccination immunity does not wane over model timescale (i.e., single pandemic wave)
%     \item Vaccination reduces infection rate by a proportion
%     \item Some constant fraction of cases are asymptomatic
%     \item Vaccination rate is constant (not necessary since we have data on exact vaccination rates)
%     \item Transmission rate is affected by current level of state policy (determined by the stringency index) and the current level of infection (corresponding to social distancing)
% \end{itemize}

% \begin{figure}[h!]
%     \centering
%     \input{diagram.tikz}v   
% \caption{State diagram of model with two different types of vaccination and strong competition. Variables are described in Table \ref{variables}}
% \label{model_structure}
% \end{figure}

% \subsection*{Further extensions}

% \begin{itemize}
%     \item Compartments will be added to test the effect of population and policy heterogeneity on viral prevalence
%     \item Postal code data for genomes would enable us to track mutations in space, and compare to small variations PHU-level closure data in a network model such as \cite{Fair_Karatayev_Anand_Bauch_2021}
    
%     \item There is a lot of genomic information not captured in high-level lineage assignments that could be used in a more 
%     detailed 
%     more of 
%     viral strain competition

%     \item Another technique that is popular for estimating epidemiological parameters, particularly in the comparison of VoC is a semi-mechanistic approach (e.g.\cite{Cauchemez_Nouvellet,Fraser_2007,Mishra_Berah_Mellan_Unwin_Vollmer_Parag_Gandy_Flaxman_Bhatt_2020,Nouvellet_Cori, Wallinga_Lipsitch_2007, Brown_Joh_Buchan_Daneman_Mishra_Patel_Day_2021})
% \end{itemize}


% \begin{table}[h!]
%     \begin{center}
%     \begin{tabular}{c|l}
%             Symbol & Description \\
%             \hline
%             \hline
%             $S$ & Susceptible, unvaccinated \\
%             $A^{(i)}$ & Asymptomatically infected with $i$th variant\\
%             $I^{(i)}$ & Infected with $i$th variant \\
%             $E^{(i)}$ & Exposed to $i$th variant \\
%             $R^{(i)}$ & Recovered from $i$th variant \\
%             $S_{V_i}$ & Susceptible, unvaccinated \\
%             $A^{(i)}_{V_i}$ & Asymptomatically infected with $i$th variant, vaccinated by $i$th vaccine\\
%             $I^{(i)}_{V_i}$ & Infected with $i$th variant, vaccinated by $i$th vaccine\\
%             $E^{(i)}_{V_i}$ & Exposed to $i$th variant, vaccinated by $i$th vaccine\\
%             $R^{(i)}_{V_i}$ & Recovered from $i$th variant, vaccinated by $i$th vaccine\\
%             $C(t, \Sigma_i I^{(i)}) $ & Function reducing transmission rate due to NPIs \\
%             $\beta_i$ & Transmission rate of $i$th variant \\
%             $\eta_i(t)$ & Vaccination rate for $i$ vaccine\\
%             $s$ & Fraction of asymptomatic \\
%             $\sigma^{(i)}$ & Inverse of latent period for $i$th variant \\
%             $\gamma^{(i)}$ & Recovery rate for $i$th variant \\
%             $\nu_{V_i}$ & Infectiousness reduction for vaccination by $i$th vaccine \\
%     \end{tabular}
%     \caption{Table of symbols for Model 1}

%     \label{variables}
%     \end{center}
% \end{table}