\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{tikz}
\usepackage{tikzit}

\usepackage{authblk}
\usepackage{blindtext}

\usepackage{url}
\usepackage{cite}
%% -------------------------------------- Declare the layers
\pgfdeclarelayer{nodelayer}
\pgfdeclarelayer{edgelayer}
\pgfsetlayers{edgelayer,nodelayer,main}
% Node styles
\tikzstyle{black circle}=[fill=white, draw=black, shape=circle, tikzit shape=circle]
\tikzstyle{Arrow}=[->]

% Edge styles
\tikzstyle{new edge style 0}=[->]

\title{Characterizing evolution pressures in Sars-CoV-2 on a broader scale}
\author[1,4]{Peter C. Jentsch}  
\author[3,5]{Finlay Maguire}
\author[1,2]{Samira Mubareka}
\affil[1]{Sunnybrook Research Institute, Toronto, Canada}
\affil[2]{University of Toronto, Toronto, Canada}
\affil[3]{Dalhousie University, Halifax, Canada}
\affil[4]{Simon Fraser University, Burnaby, Canada}
\affil[5]{Shared Hospital Laboratory, Toronto, Canada}
\date{\today}                     %% if you don't need date to appear
\setcounter{Maxaffil}{0}
\renewcommand\Affilfont{\itshape\small}

\begin{document}


\maketitle

\section{Background}
Previous work on these models do not use data to estimate parameters, focusing on broad characterization of dynamics with numerical or analytical approaches. However, genomic data does include a huge amount of admittedly very noisy information. There is some recent work on parameter estimation by comparing simulated phylogeny with observed phylogeny using a suite of summary statistics for tree structures \cite{danesh2021quantifying,leventhal2012inferring,saulnier2017inferring}, but these models do not explicitly include genomic structure, which might provide additional inferential power. 

Placing individual samples into the collapsed strain space could be done by using neutralizing antibody studies. The neutralizing response has been characterized for most variants of concern with respect to monoclonal antibodies, convalescent plasma, and plasma from vaccinated persons \cite{stanford2020}. To an even higher level of detail. Starr et al. characterized polyclonal antibody binding over all mutations of the Receptor Binding Domain (RBD) \cite{starr2020deep}. If the neutralizing response is known, we assign strains positions such that the distance between any two strains is equal to that neutralizing response. If the response is not known, we just assume that the position is close to the most recent ancestor for which the response is known (likely a VoC). This approach is especially appropriate for recurrent mutations, since it is unlikely that a mutation that has occurred many times in the viral history but has not evolved further confers a significant advantage. The immune-space position of vaccinate-induced immunity can also be computed this way, although we will still need to make significant assumptions on their geometry. It might be best to assume a simple symmetric shape with a centre given by neutralization results. 

Mirroring the work on mapping the strain space for influenza \cite{lapedesGeometryShapeSpace2001, smithMappingAntigenicGenetic2004, cai2010computational} computing this embedding is a metric multidimensional scaling problem. This method has been applied to the antigenic space of the Sars-CoV-2 as well \cite{millerAntigenicSpaceFramework2021, wilksMappingSARSCoV2Antigenic2022, van2022mapping}. 


\section{Methods}


\bibliographystyle{plain}
\bibliography{ref.bib}

\end{document}  


% \section{Model}
% \label{model}
% \begin{itemize}
%     \item The structure of the model is Susceptible-Exposed-Infected-Asymptomatic-Recovered, outlined in Figure \ref{model_structure}.
%     \item The mRNA vaccine lowers transmission rates and increases the proportion of asymptomatic infections compared to the AstraZeneca vaccine
%     \item Focus on infections in two lineages
%     \item Each lineage has different infection rates, recovery times, and latent periods
%     \item Co-infection negligible 
%     \item Time scale short enough that recovery from either infection grants immunity
%     \item Strains interact in the model only in that they compete for hosts
%     \item Vaccination immunity does not wane over model timescale (i.e., single pandemic wave)
%     \item Vaccination reduces infection rate by a proportion
%     \item Some constant fraction of cases are asymptomatic
%     \item Vaccination rate is constant (not necessary since we have data on exact vaccination rates)
%     \item Transmission rate is affected by current level of state policy (determined by the stringency index) and the current level of infection (corresponding to social distancing)
% \end{itemize}

% \begin{figure}[h!]
%     \centering
%     \input{diagram.tikz}v   
% \caption{State diagram of model with two different types of vaccination and strong competition. Variables are described in Table \ref{variables}}
% \label{model_structure}
% \end{figure}

% \subsection*{Further extensions}

% \begin{itemize}
%     \item Compartments will be added to test the effect of population and policy heterogeneity on viral prevalence
%     \item Postal code data for genomes would enable us to track mutations in space, and compare to small variations PHU-level closure data in a network model such as \cite{Fair_Karatayev_Anand_Bauch_2021}
    
%     \item There is a lot of genomic information not captured in high-level lineage assignments that could be used in a more 
%     detailed 
%     more of 
%     viral strain competition

%     \item Another technique that is popular for estimating epidemiological parameters, particularly in the comparison of VoC is a semi-mechanistic approach (e.g.\cite{Cauchemez_Nouvellet,Fraser_2007,Mishra_Berah_Mellan_Unwin_Vollmer_Parag_Gandy_Flaxman_Bhatt_2020,Nouvellet_Cori, Wallinga_Lipsitch_2007, Brown_Joh_Buchan_Daneman_Mishra_Patel_Day_2021})
% \end{itemize}


% \begin{table}[h!]
%     \begin{center}
%     \begin{tabular}{c|l}
%             Symbol & Description \\
%             \hline
%             \hline
%             $S$ & Susceptible, unvaccinated \\
%             $A^{(i)}$ & Asymptomatically infected with $i$th variant\\
%             $I^{(i)}$ & Infected with $i$th variant \\
%             $E^{(i)}$ & Exposed to $i$th variant \\
%             $R^{(i)}$ & Recovered from $i$th variant \\
%             $S_{V_i}$ & Susceptible, unvaccinated \\
%             $A^{(i)}_{V_i}$ & Asymptomatically infected with $i$th variant, vaccinated by $i$th vaccine\\
%             $I^{(i)}_{V_i}$ & Infected with $i$th variant, vaccinated by $i$th vaccine\\
%             $E^{(i)}_{V_i}$ & Exposed to $i$th variant, vaccinated by $i$th vaccine\\
%             $R^{(i)}_{V_i}$ & Recovered from $i$th variant, vaccinated by $i$th vaccine\\
%             $C(t, \Sigma_i I^{(i)}) $ & Function reducing transmission rate due to NPIs \\
%             $\beta_i$ & Transmission rate of $i$th variant \\
%             $\eta_i(t)$ & Vaccination rate for $i$ vaccine\\
%             $s$ & Fraction of asymptomatic \\
%             $\sigma^{(i)}$ & Inverse of latent period for $i$th variant \\
%             $\gamma^{(i)}$ & Recovery rate for $i$th variant \\
%             $\nu_{V_i}$ & Infectiousness reduction for vaccination by $i$th vaccine \\
%     \end{tabular}
%     \caption{Table of symbols for Model 1}

%     \label{variables}
%     \end{center}
% \end{table}